astropy/table/table.py

```diff
@@ -801,58 +828,97 @@ def _init_from_list(self, data, names, dtype, n_cols, copy):
-                                       copy=copy, copy_indices=self._init_indices)
-
-            else:
-                raise ValueError('Elements in list initialization must be '
-                                 'either Column or list-like')
+            col = self._convert_data_to_col(col, copy, def_name, dtype, name)
 
             cols.append(col)
 
         self._init_from_cols(cols)
 
+    def _convert_data_to_col(self, data, copy=True, def_name=None, dtype=None, name=None):
+        """
+        Convert any allowed sequence data ``col`` to a column object that can be used
+        directly in the self.columns dict.  This could be a Column, MaskedColumn,
+        or mixin column.
+
+        The final column name is determined by::
+
+          name or data.info.name or def_name
+
+        If ``data`` has no ``info`` then ``name = name or def_name``.
+
+        Parameters
+        ----------
+        data : object (column-like sequence)
+            Input column data
+        copy : bool
+            Make a copy
+        def_name : str
+            Default name
+        dtype : np.dtype or None
+            Data dtype
+        name : str or None
+            Column name
+
+        Returns
+        -------
+        col : Column, MaskedColumn, mixin-column type
+            Object that can be used as a column in self
+        """
+        # Structured ndarray gets viewed as a mixin unless already a valid
+        # mixin class
+        if (isinstance(data, np.ndarray) and len(data.dtype) > 1 and
+                not self._add_as_mixin_column(data)):
+            data = data.view(NdarrayMixin)
+
+        # Get the final column name using precedence.  Some objects may not
+        # have an info attribute.
+        if hasattr(data, 'info'):
+            name = name or data.info.name or def_name
+        else:
+            name = name or def_name
+
+        if isinstance(data, Column):
+            # If col is a subclass of self.ColumnClass, then "upgrade" to ColumnClass,
+            # otherwise just use the original class.  The most common case is a
+            # table with masked=True and ColumnClass=MaskedColumn.  Then a Column
+            # gets upgraded to MaskedColumn, but the converse (pre-4.0) behavior
+            # of downgrading from MaskedColumn to Column (for non-masked table)
+            # does not happen.
+            if issubclass(self.ColumnClass, data.__class__):
+                col_cls = self.ColumnClass
+            else:
+                col_cls = data.__class__
+
+            col = col_cls(name=name, data=data, dtype=dtype,
+                          copy=copy, copy_indices=self._init_indices)
+
+        elif self._add_as_mixin_column(data):
+            # Copy the mixin column attributes if they exist since the copy below
+            # may not get this attribute.
+            col = col_copy(data, copy_indices=self._init_indices) if copy else data
+
+            col.info.name = name
```

Comment:
If we add `return col` here, then all the other parts of the long `if` statement could fall through to a shared
```
            col = col_cls(name=name, data=data, dtype=dtype,
                          copy=copy, copy_indices=self._init_indices)
```
at the end.

Reply:
Excellent, done.